design: Node Image and Kubernetes Version Controls for NAP (#1756)
Azure/karpenter-provider-azure: designs/0014-node-image-and-k8s-version-controls.md

> # Node Image and Kubernetes Version Controls for NAP

Reviewer: I like this title more than the PR title "Node Image Rollback", because it is more than just rollback, right?

Maybe correct the PR title to match this title? It feels more accurate to me.

Author: yep forgot to change the pr title, my bad

**Author:** @rakechill

**Last updated:** Jun 26, 2026

**Status:** Proposed

**Related issues:** [Azure/karpenter-provider-azure#1220](https://github.com/Azure/karpenter-provider-azure/issues/1220), [Azure/karpenter-provider-azure#1355](https://github.com/Azure/karpenter-provider-azure/issues/1355)

> Note: The issue title says "Add support for setting ImageID for nodeClass", but the problem statement is primarily asking for rollback/pinning capability to recover from bad image releases. This rollback design addresses the bad-release recovery use case. It does not address the literal title request to set arbitrary ImageID; that is likely a future AKS workstream, and Karpenter should not set that standard prematurely.

Reviewer: Also to prevent "unplanned" uptake of new images until user has time to verify them

Author: TODO: add this context to the statement

8. [Drift and Provisioning Behavior](#drift-and-provisioning-behavior)
   - [Existing nodes](#existing-nodes)
   - [New scale-ups](#new-scale-ups)
   - [Drift trigger choice](#drift-trigger-choice)
   - [Kubernetes version drift](#kubernetes-version-drift)
9. [Decision Notes](#decision-notes)
10. [Open Questions](#open-questions)
11. [Out of Scope Follow-up Designs](#out-of-scope-follow-up-designs)
12. [Production Readiness](#production-readiness)
13. [References](#references)

## Overview

This document proposes version control capabilities for AKS Node Auto Provisioning (NAP) in Karpenter, including node image version pinning, rollback, and Kubernetes version drift control. These capabilities are aligned with AKS agent pool rollback and upgrade semantics.

The intent is to let customers control which node image version their NAP nodes run on: either by pinning to the current version to prevent automatic upgrades, or by rolling back to the previously used version to recover from a bad image release. Additionally, customers can specify a target Kubernetes version on the NodeClass to decouple node Kubernetes version drift detection from the control plane's current version.

> Related but separate workstreams are:

Reviewer: minor: YOu also call these out as non-goals. I think it's maybe fine/good to have them in both places so just confirming that is the intent. It's a bit of duplicate information but probably reads well/ok?

Author: yeah was attempting to immediately limit the scope of this design, but also call out that design will be needed for these soon, too.

1. Long-duration arbitrary node image pinning beyond current and previously used versions.
2. Prepared image spec support using a full resource ID that maps to an AKS API prepared image field.
3. Future image version selectors for filtering and ranking available image versions.

These workstreams require independent design and are out of scope here.

## Background

AKS node images are VHD-based image versions that are updated frequently. Today, NAP users configure selector-style image controls in spec (for example image family), while resolved concrete images are surfaced in status.

### Node image version updates

> When a new node image version becomes available, Karpenter automatically picks it up and surfaces it in `status.images`. Existing nodes running the prior image are then considered drifted and replaced according to normal disruption controls, subject to maintenance windows and disruption budgets. Customers have no mechanism today to opt out of this process, defer it, or manually trigger it at a time of their choosing: the upgrade happens whenever the disruption controller determines it is safe to act.

Reviewer: I mean they can use maintenance windows but they are a blunt instrument. I do think this statement is incorrect though:

> Customers have no mechanism today to opt out of this process, defer it, or manually trigger it at a time of their choosing:

They have no way to opt out of manually trigger it, but they can defer it with maintenanceWindows.

Author: Yeah "no way" is probably inaccurate here. Something more along the lines of -- insufficient fine-grained controls + no support for manual upgrades without strict + explicit maintenance window configuration.

I'll think of a better way to phrase this...

### Kubernetes version upgrades

> NAP node Kubernetes version upgrades behave differently from AKS managed agent pools. When the cluster control plane is upgraded to a new Kubernetes version, Karpenter immediately recognizes the version delta and marks affected nodes as drifted. Replacement is then driven by the standard disruption flow, respecting maintenance windows and disruption budgets, but customers cannot separately stage or defer the node k8s version upgrade the way they can with AKS agent pool upgrade controls.

Reviewer: minor note: It's not _quite_ immediate there is a polling interval - maybe say quickly instead of immediately?

Author: Fair

TODO: update this language to quickly/soon after.

Additionally, a Kubernetes version upgrade also triggers a node image version refresh: nodes are replaced with the latest node image compatible with the new Kubernetes version. This means a control plane upgrade causes both a k8s version change and a node image change on NAP nodes simultaneously, neither of which is individually controllable today.

### Gap this design addresses

Current NAP behavior has no rollback affordance and no version control surface:

> 1. No spec field lets customers request rollback or pin to a specific node image version.

Reviewer: ```suggestion
1. No spec field lets customers request rollback or control node image version rollout manually.
```

Pin has certain connotations which maybe we want to avoid (even though the API I think you're proposing is really "pin-lite")

2. No spec field lets customers decouple their NAP nodes' Kubernetes version from the control plane's current version.

> 3. No status field preserves the previous image set as a first-class rollback target.

Reviewer: Nice callout, agree this would be good.

> 4. Drift logic replaces nodes that are not on the current status image set, but does not model rollback intent.

Reviewer: Not sure I am following what this means - can you clarify?

Author: I think this statement is saying nodes will get drifted if they aren't using the image from `status.images[]`. And in a previous iteration of the design, if a customer requests rollback, I was planning on keeping the same latest images in the `status.images[]`, but overwriting the imageVersion portion during drift/scale up.

So I think in that case, `status.images[]` would be unchanged and thus drift/image resolution logic would have to additionally be updated to account for rollback intent + "overwrite" the `status.images[]`.

Basically, I think this is a confusing statement + not that useful for this section.

TODO: remove

> AKS RP rollback semantics rely on a recently-used allowlist. NAP currently has no equivalent mechanism.

Reviewer: Isn't this the same (or almost the same) as bullet point 3 above? Can it be slipstreamed into there somehow?

Author: TODO: reference that the recently used versions field will model this + remove the separate statement.

## Goals

1. Provide a `nodeImageVersion` spec field that allows customers to pin to their current node image version or roll back to the previously used version.
2. Preserve recently-used version state durably in AKSNodeClass status, including the Kubernetes version paired with the previously used node image.

> 3. Keep rollback and pinning operationally safe and predictable. Pinning is limited to the current and previously used versions only.

Reviewer: Call out that this is the same rule that AKS RP itself uses.

Author: TODO: mention this is due to AKS parity.

4. Provide a `kubernetesVersion` spec field that allows customers to specify the desired Kubernetes version for drift detection, decoupling node k8s version drift from the cluster control plane version.
5. Surface both the current node image version and the recently used node image version (with its paired Kubernetes version) in AKSNodeClass status.
6. Keep the API design extensible for future image version selectors.

## Non-Goals

1. Implement long-duration or arbitrary image pinning beyond the current and previously used versions.
2. Introduce prepared image spec by full resource ID.
3. Guarantee rollback to arbitrary historical versions.
4. Block provisioning scale-up while rollback state is being evaluated.

## API Changes

### API Field Grouping and Wrapper Name

`nodeImageVersion` and `kubernetesVersion` should be grouped under a dedicated sub-section rather than placed at the top level of spec. Top-level placement would crowd the spec namespace and leave no clean home for future image version selectors.

The proposed shape is:

```yaml
spec:
  <wrapper>:

>     kubernetesVersion: "1.32"

Reviewer: One interesting question... if I set 1.32 here am I saying I _do_ want disruption/drift if the upstream k8s APIServer changes minor version? Or that I want the latest minor version?

This is the [alias minor version](https://learn.microsoft.com/en-us/azure/aks/supported-kubernetes-versions?tabs=azure-cli#alias-minor-version) feature of AKS and it does maybe make sense to support something very similar in Karpenter... though we could defer support to later if you want to keep the scope a bit smaller (if so, please file an item and assign either yourself or me and we can take a look at doing it later).

Basically the question is can users say 1.32 here and what exactly does that mean? It could mean:
1. I want 1.32 and patch me to the latest 1.32.x that exists regardless of controlplane version (could end up with controlplane 1.32.2 and nodes 1.32.4)
2. I want 1.32, and patch me to the latest 1.32 that the control plane is also using. So when controlPlane goes 1.32.2->1.32.4, so will the nodes.
3. I want 1.32.x where .x is the latest at the time I set this, and then don't ever drift them again

I do not favor 3. I like 1 or 2 but I think 2 is safer/easier so tend towards that. If we do not want to allow this case currently I would strongly suggest we just require the full 3 x.y.z rather than support x.y. (then as mentioned above file an item)

Author: There are specific guidelines around this in AKS. I _think_ I covered this in my design, but if I didn't then it's in some notes that I need to reflect here.

TODO: validate that this is either covered in the design or bring the context in. determine if it should/could be punted to a future design instead.

Author: Required explicit `major.minor.patch` to reduce difficult questions around drift.

    nodeImageVersion: "202601.15.0"

>     # future: selectors: [...]

Reviewer: If we're thinking selectors in the future, I am not sure the proposed API is fully future-proof. The reason being, ideally you would have kubernetesVersion and nodeImageVersion pinned (select specific image/version) as here, as a type of selector.

Refer to what AWS does [here=(https://github.com/aws/karpenter-provider-aws/blob/main/pkg/apis/v1/ec2nodeclass.go#L76C21-L76C37) for example.

That _could_ suggest either:
```
spec:
  kubernernetesVersionSelectors: 
    - version: "1.32"
  nodeImageVersionSelectors:
    - version: "202601.15.0"
```

or

```
spec:
  <wrapper>: 
    kubernernetesVersionSelectors: 
      - version: "1.32"
    nodeImageVersionSelectors:
      - version: "202601.15.0"
```

I _tend_ to think the first is better as each version/set of versions has its own area already and there's probably less need to co-locate k8sversion + nodeImageVersion under the same object at that point.

Reviewer: I think this then extends to something like:
```
    kubernernetesVersionSelectors: 
      - tags:
        foo: bar
```

if we wanted? Would be good to get @tallaxes to confirm.

Author: I am wondering if we want a separate(?) spec field/section to cover our current image fields that are essentially image selectors themselves. And then we keep _version_ separate from those considerations

Basically migrating to:
```
imageSelectors:
- imageFamily: Ubuntu
- fipsMode: true
- tags:
   foo: bar
...
```

```

The open question is the name of the wrapper. Three candidates:

| Name | Pros | Cons |
|---|---|---|
| `manualUpgrade` | Short. Clearly signals opt-out of automatic version management. "Manual" is the natural opposite of "auto". | "Upgrade" implies forward movement — rolling back or pinning at current is not strictly an upgrade. |
| `upgradeControl` | Directionally neutral — covers upgrading, pinning, and rolling back without implying movement. Reads naturally in a Kubernetes API context. | Slightly more abstract. Does not immediately convey the opt-out-of-auto framing. |
| `manualVersionControl` | Accurate description of what the fields do. | "Version control" carries strong source-control connotations (git, svn). Longest of the three. |

> **Recommendation:** `upgradeControl` — semantically precise, direction-neutral, and fits the Kubernetes API style. `manualUpgrade` is a strong second if the opt-out-of-auto signal is prioritized over directional neutrality.

Reviewer: I don't love `upgradeControl` because it's a bit "action-y", and we're talking about a spec field here which is about the steady state, not the dynamic/action-state of upgrading.

`versionControl` or `versions` or similar I prefer slightly more, though see my comments/questions above about selectors and how that might change things.

Author: I don't _love_ `versionControl` because I immediately think git :/

but it is basically accurate aside from that.

I agree that "upgrade" makes it feel like we are forward progressing which is inaccurate in some cases + we could just be adjusting the version we're already on...

Author: `versionSelection` is what I'm thinking now

**AKS version skew constraint:**

AKS enforces a version skew policy between the cluster control plane and node pools. For clusters running Kubernetes 1.28 and later, the node pool Kubernetes version must be within **three minor versions** of the control plane (N-3). The node pool version also cannot be greater than the control plane version. This is documented in the [AKS supported Kubernetes versions FAQ](https://learn.microsoft.com/en-us/azure/aks/supported-kubernetes-versions#what-is-the-allowed-difference-in-versions-between-the-control-plane-and-node-pools).

When `spec.upgradeControl.kubernetesVersion` is set, Karpenter must validate that the specified version satisfies this constraint relative to the cluster control plane version before provisioning or during drift evaluation:

1. The specified version must not be greater than the control plane version.
2. The specified version must not be more than three minor versions behind the control plane version (for Kubernetes 1.28+).

> Violations should be surfaced as a condition on the NodeClass rather than silently accepted, since AKS will reject the machine creation at the RP level if the skew constraint is violated.

Reviewer: open questions or FAQ section lacking discussion around what happens when this is set to something 3 versions ago and then controlPlane upgrades. Today, AKS RP will (I think) block the controlplane upgrade if that happens. We should (probably?) integrate similar logic in AKS RP for Karpenter nodes or just all nodes in general? This isn't an issue today because we don't really allow locking node images in karpenter today, but now that we will I wonder if version validation in AKS RP should take managed NAP nodes into account in addition to AgentPools and block controlPlane upgrade if it would end up >3 versions ahead.

Reviewer: We already do something _slightly_ similar in the deprecated API usages checker in AKS RP. There are some risks with going to check apiserver directly (it can be down in various cases that are "expected"), but a check that's at least best-effort and fails-open in AKS RP probably makes sense.

Author: yeah I think that makes sense post-Machine API migration. I think we also may need to ensure an LTS version isn't selected. I'll add this.

TODO: 
* update Open Questions section to include interaction between controlPlane upgrades (RP validation) and nodeclass k8s version.
* update this section to cover LTS version constraints

**Version granularity:** Whether this field accepts patch versions (e.g. `1.32.5`) or only minor versions (e.g. `1.32`) is an open question. AKS agent pool behavior needs to be researched before finalizing accepted values. See Open Questions.

### nodeImageVersion Spec Field

The `nodeImageVersion` spec field is the unified customer surface for both node image pinning and rollback. It determines which node image release version Karpenter targets for new provisioning and drift evaluation.

**Semantics:**

1. **Unset (default):** Karpenter always resolves and uses the latest available node image version. This is the existing NAP behavior.

> 2. **Set to `status.latestImageVersion`:** Karpenter pins to the latest resolved version. New nodes are provisioned on that version. Automatic node image upgrades are paused — if a newer version becomes available, nodes will not drift until the customer updates or clears the pin.

Reviewer: Is this _really_ latest or does it honor MW similar to how images does today and only update when MW is open?

Author: So the latest image versions are listed/fetched during EVERY nodeclass reconcile w/ requeue after set to 5min. We just don't override the images[] until a maintenance window is open. I was thinking latestImageVersion field would be set upon each reconcile. But I need to think a bit more on the edge cases here. 

Some scenarios...

(1) Customer has no pinning, but has maintenance windows confugured
- Version A is the previous version which can be seen in recentlyUsedVersions struct.
- Version B is the image used in images[] and was the latest in the last MW. The customer DID NOT pin/manually select.
- Version C is the latest image version and is set in latestImageVersion

---> When MW opens, there are no pins, so SIG.List() is called + the images[] get overridden with Version C.

(2) Customer has pinned their version, but has maintenance windows confugured
- Version A is the previous version which can be seen in recentlyUsedVersions struct.
- Version B is the image used in images[] and was pinned by the customer.
- Version C is the latest image version and is set in latestImageVersion

---> When MW opens, the latest image version is still captured (and can be pinned to), but images[] is not overridden because Version B is pinned.

(3) Customer has not pinned their version, but tries to pin before MW opens and after there's a new latestImageVersion
- Version A is the previous version which can be seen in recentlyUsedVersions struct.
- Version B is the image used in images[], but will be reset when the MW opens.
- Version C is the latest image version and is set in latestImageVersion

Customer tries to pin to the image version in images[]. Should the expected behavior be:
- Rejected, because version != recentlyUsedVersion && version != latestImageVersion
- Accepted, because version is currently used in images[]

I think it should be Rejected, because we currently just want customers to be able to pin to latest or their previous version...
WDYT?

Author: Importantly, if they want to pin to Version B **after** picking up Version C, they still can. Why? Because this will be the recentlyUsedVersion after the bump to Version C.

Reviewer: > I think it should be Rejected, because we currently just want customers to be able to pin to latest or their previous version...

This doesn't actually quite match the AKS behavior though, does it? Since for AKS pinning is basically the default (we don't roll nodes unless user takes an explicit action or configures autoupgrade).

Consider the case that we have an outage and a bad node image version is published. The user has currently (due to MW or other reasons) not yet detected the new node image version in `images[]`, but it _has_ been detected in `latestImageVersion` (because that gets detected very quickly and updated). They want to prevent uptake of that new version as they got an AzComms that it is bad. How can they do that? They can extend their MW, but they wouldn't actaully be able to pin - BUT, they could upgrade to the bad version (thereby pushing current good image into recentlyUsedVersions). Now they're on the bad image, BUT they can immediately roll back to the `recentlyUsedVersions` version (right?)

So I think we want to allow pinning to:
1. The previous version (recentlyUsedVersions)
2. The current version
3. The latest version

Because it seems to me otherwise you get some weird artifacts where you _cant_ pin to a particular version (version A, but if you upgrade to a _different_ version then you _can_ pin to that version A, which seems real weird to me.

Reviewer: but ack that:
> Is this really latest or does it honor MW similar to how images does today and only update when MW is open?

It is really latest, because it is the key off which CEL validation is triggered that allows you to pin/upgrade to the latest, so it can't honor MW.

Author: > Consider the case that we have an outage and a bad node image version is published...

I did not think of rollback as being a proactive, rather than reactive measure in many cases. But that makes a lot of sense. Especially if you know your MW will pick it up soon, but your nodes have not yet been impacted. 

So, agreed we should allow all three (in the case that current != latest):
1. The previous version (recentlyUsedVersions)
2. The current version
3. The latest version

2. A customer could silently roll back to a version they did not intend.
3. If Karpenter later stores multiple previous image versions, a boolean becomes ambiguous without additional API to specify which entry to use.
4. The explicit version string approach mirrors AKS RP semantics, where `nodeImageVersion` is always a concrete version value.

### AKSNodeClass status: latestImageVersion and recentlyUsedVersions

Add a new status section that mirrors the AKS RP recently-used rollback model:

```go
type RecentlyUsedVersions struct {
	// timestampUsed is the time this node image version was last active,
	// recorded for observability when Karpenter moves status.images forward.
	TimestampUsed metav1.Time `json:"timestampUsed,omitempty"`

	// imageVersion is the AKS node image release version suffix used for rollback,
	// e.g. "202601.15.0".
	ImageVersion string `json:"imageVersion,omitempty"`

	// kubernetesVersion is the Kubernetes version paired with this image,
	// e.g. "1.32.5".
	KubernetesVersion string `json:"kubernetesVersion,omitempty"`
}
```

```go
type AKSNodeClassStatus struct {
	// existing fields...

> 	Images               []NodeImage             `json:"images,omitempty"`

Reviewer: minor: This already exists, but it's not clear based on the shape you have defined here (shouldn't this be in existing fields... ?) If you're calling this out explicitly just because it's important/used in this design (which makes sense to me), maybe put `// existing` alongside it?

Author: TODO: instead of having `// existing fields...`, remove that line and add a clear `// existing` alongside `Images`

> 	LatestImageVersion   string                  `json:"latestImageVersion,omitempty"`

Reviewer: minor: Missing things like `// +optional` and also standard `//` comments explaining what each means, which would be helpful in this context (and in the code that this will eventually become).

Author: TODO: add full struct definition w/ +optional + explanation as will be added to https://github.com/Azure/karpenter-provider-azure/blob/main/pkg/apis/v1beta1/aksnodeclass.go

> 	LatestImageVersion   string                  `json:"latestImageVersion,omitempty"`

Reviewer: How does classifying `latestImageVersion` work when _technically_ there can be multiple latests for different images? (it is allowed for in the `status.images` field even if it doesn't happen in practice today -- though to be fair I am not actually 100% sure it _never_ happens I vaguely recall Charlie getting them to admit it had happened in the past for some very specific cases).

I wonder if we want/need to cater for that in our API design though just on the off chance it happens, but in a way where the default/current behavior is as you state here (there's only ever one image). Then at least if we ever end up in a situation where there is or _will be_ multiple images we have an out.

My current somewhat brainstorm-y idea is:
```
status:
  images: // existing
    - id: /subscriptions/10945678-1234-1234-1234-123456789012/resourceGroups/AKS-Ubuntu/providers/Microsoft.Compute/galleries/AKSUbuntu/images/2204gen2containerd/versions/2022.10.03
      requirements: 
        - <various requirements>
  latestImages: // new
    - version: 2022.10.03
```

but the actual format of `status.latestImages` would support a `requirements` field too, and so in a case where we had multiple latests we could fill out requirements for each (basically instead of a single entry with no requirements which means it matches everything, it could be 2 entries with mutually exclusive requirements). We don't need to do this/support this now but it would mean we _could_ in the future? I am not sure if requirements is the best way to slice/dice or if there is a better way (tags or something?) requirements just matches what `images` uses so might be nice...

Then it would beg the question how the heck do you even select a different image version for different node types, which is obviously somewhat painful/complicated and muddles the ux, _but_ maybe it already fits in well with the `selectors` approach discussed [here](https://github.com/Azure/karpenter-provider-azure/pull/1756/changes#r3669691420), and we could have a selector with similar requirements section (or based on tags/etc) that lines up with this and basically validate that when there are 2x different versions in latestImages, there must _also_ be matching selector in the selectors so that you're specifically selecting for one or the other (since you can't set the same version for both).

Reviewer: cc @tallaxes because this is kinda a mess but it's also what we already technically support even if it doesn't happen regularly in practice. Maybe we need to confirm with @cameronmeissner and co if this can happen (I am _pretty_ sure it can).

Author: > How does classifying latestImageVersion work when technically there can be multiple latests for different images?

Interesting ok. Maybe need to chat with upgrade/node folks about this.

Author: more info in the other thread. i think we're ok here for now
https://github.com/Azure/karpenter-provider-azure/pull/1756#discussion_r3707118226

> 	RecentlyUsedVersions *RecentlyUsedVersions   `json:"recentlyUsedVersions,omitempty"`

Reviewer: minor: This field reads like a collection but is not. maybe we want it to be a collection though to extend to further back than n-1? My recollection was that was discussed in Indu's API too and supported there, see [here](https://github.com/Azure/azure-rest-api-specs/blob/main/specification/containerservice/resource-manager/Microsoft.ContainerService/aks/preview/2026-05-02-preview/managedClusters.json#L7809) -- it's an array not a singleton.

Reviewer: Then we just by convention only ever keep 1 item in here and if we ever want to extend to 2, 3, etc we can easily without breaking change.

Author: TODO: update the field type here to properly reflect intent

}
```

Semantics:

1. recentlyUsedVersions captures the immediate prior node image release version suffix and Kubernetes version pair before status.images advances.
2. recentlyUsedVersions.timestampUsed records when the previous version was last active, for observability.
3. latestImageVersion always reflects the latest resolved image version suffix, regardless of whether rollback or pinning is active. It is updated on every reconcile pass, even when status.images is overwritten with a rolled-back or pinned version.
4. Only one recently-used entry is retained, matching AKS RP semantics.
5. If Karpenter stores multiple previous image versions in the future, rollback UX must specify how Karpenter chooses which previous version to use.

## Reconciliation Design

### Snapshot point

NodeImageReconciler in images.go updates `status.latestImageVersion` on every reconcile pass. A snapshot into `status.recentlyUsedVersions` is taken whenever the **effective image version changes**. The three triggers are:

- **Gallery advance:** `status.latestImageVersion` moves to a newer version while `nodeImageVersion` is unset.

> - **Customer sets `nodeImageVersion`:** effective version changes from what was in `status.images` to the newly requested value.

Reviewer: maybe this is just saying the same thing as what you're saying, but doesn't this boil down to:
"Any time the images field is updated"?

Would it be easier to talk about it in that way? basically `status.images` is what images we are using. Any time that changes, we should re-evaluate `recentlyUsedVersions`, right?

Author: Yes, correct.

- **Customer unsets `nodeImageVersion`:** effective version changes from the pinned version back to `status.latestImageVersion`.

In all cases the snapshot captures the version being left, so `status.recentlyUsedVersions.imageVersion` always represents "the previous effective version before the current one." `status.recentlyUsedVersions.kubernetesVersion` reflects `spec.upgradeControl.kubernetesVersion` if set, otherwise the cluster control plane version.

### Rollback path

When `spec.upgradeControl.nodeImageVersion` is set to the previously used version:

1. Validate `status.recentlyUsedVersions` exists and has an `imageVersion`.
2. Validate the requested `nodeImageVersion` matches `status.recentlyUsedVersions.imageVersion`.

> 3. Validate the currently desired Kubernetes version is compatible with `status.recentlyUsedVersions.kubernetesVersion`. The desired Kubernetes version is `spec.upgradeControl.kubernetesVersion` if set, otherwise the cluster control plane version.

Reviewer: If the version is coming from cluster control plane version (default) then recentlyUsedVersions doesn't matter at all, does it?

Author: recentlyUsedVersions = [
         {k8s=1.32, nodeImage=2022.10.13}
]
nodeImageVersion = 2022.10.20
k8sVersion = `unset`

If k8s version is upgraded to 1.33 while on `2022.10.20`, but customer tries to set nodeImageVersion to `2022.10.13`, we would need to reject this.

maybe i'm confused by your question tho

4. If valid, set the effective target image release version suffix to `status.recentlyUsedVersions.imageVersion`.
5. Apply the rollback image version suffix to `status.images` per the implementation decision (Option 2).

### Image selection during rollback

status.images is already a list because a single AKSNodeClass can expose multiple compatible image definitions. For example, the same NodeClass may publish Gen2 amd64, Gen1 amd64, and Gen2 arm64 images, each with requirements that determine which instance types can use it.

Rollback should preserve this model:

1. Karpenter resolves the normal goal image list from the AKSNodeClass, Kubernetes version, FIPS mode, SIG/CIG mode, and supported image definitions.
2. Each resolved image keeps its existing requirements. Node launch continues to choose the first image whose requirements are compatible with the selected instance type.

> 3. Rollback does not choose one image from status.images globally. Instead, it rewrites every resolved image ID to the requested release version suffix.

Reviewer: As mentioned [above](https://github.com/Azure/karpenter-provider-azure/pull/1756/changes#r3670115050) this is risky because it assumes the versions are always the same. I think we need some sort of "applies to"/requirements logic... possibly integrated into spec too (I was positing just status above).

Again we don't need to ever _set_ it for this initial implementation, but I think it might need to be there/be designed for.

Author: As far as I can tell from AgentBaker + AKS RP, it seems that security patches + releases are usually applied to ALL architecture variants within an image family, but are not guaranteed to be applied to ALL image families.

For this reason, I think this is safe wrt our current AKSNodeclass. I'll reach out to Cameron to confirm.

Author: Talked to Cameron. At the moment, this isn't something that AKS node lifecycle allows (e.g. only release an update for a subset of flavors/variants if it fails for one), but it might be something that happens in the future.

In that case, I suggest we have a protective measure that:
* checks if the image version exists for all of the compatible variants
* only populates images[] with compatible variants + maybe adds a "warning" status message to say some variants aren't available with pinned image
* only hard fails if ALL image variants don't have this image version.

Reviewer: I think I am OK with this for now, as long as node sig is saying this is not really supposed to happen today.

4. To rewrite an image ID, Karpenter strips the existing `/versions/<current>` suffix and appends `/versions/<rollback imageVersion>`.

Example:

```text
normal goal image:
/CommunityGalleries/.../images/2204gen2containerd/versions/202607.15.0

recentlyUsedVersions.imageVersion:
202606.08.1

rollback goal image:
/CommunityGalleries/.../images/2204gen2containerd/versions/202606.08.1
```

The same suffix rewrite is applied independently to each resolved image definition. This lets multiple NodePools share one AKSNodeClass while still rolling back to the image variant selected by each NodePool's instance type requirements.

Before using rolled-back images, Karpenter should verify that the reconstructed image version exists for each resolved image definition. If any required image definition does not have the requested release suffix, rollback should fail with a clear condition rather than using an invalid image ID.

> ### Implementation options for applying rollback

Reviewer: I stopped here, will have some more AI-generated comments once I validate them and probably some impl notes, but I think above comments enough to talk about for now

The intent is to let customers control which node image version their NAP nodes run on: either by pinning to the current version to prevent automatic upgrades, or by rolling back to the previously used version to recover from a bad image release. Additionally, customers can specify a target Kubernetes version on the NodeClass to decouple node Kubernetes version drift detection from the control plane's current version.

Related but separate workstreams are:

1. Long-duration arbitrary node image pinning beyond current and previously used versions.
2. Prepared image spec support using a full resource ID that maps to an AKS API prepared image field.
3. Future image version selectors for filtering and ranking available image versions.

These workstreams require independent design and are out of scope here.

## Background

AKS node images are VHD-based image versions that are updated frequently. Today, NAP users configure selector-style image controls in spec (for example image family), while resolved concrete images are surfaced in status.

### Node image version updates

> Today, the NodeImageReconciler resolves the latest gallery images on every AKSNodeClass event and on a 5-minute requeue. However, it does not always immediately publish those latest versions into `status.images`. If `ImagesReady` is false, it applies latest immediately. Otherwise, it only moves existing image definitions forward when the maintenance window is open; outside that window it preserves the existing versions and only adds newly available SKUs. If no maintenance window is configured for the node OS upgrade schedule, the current behavior fails open and applies latest.

Author: TODO: check code ref for this. when wouldn't they be ready?

Author: when there are no images available

> Existing nodes are considered drifted and replaced according to normal disruption controls once the effective image set in `status.images` moves forward. Customers have no suggested mechanism to opt out of this process, defer it beyond maintenance window shaping, or manually trigger it at a time of their choosing.

Reviewer: ```suggestion
Existing nodes are considered drifted and replaced according to normal disruption controls once the effective image set in `status.images` changes. Customers have no suggested mechanism to opt out of this process, defer it beyond maintenance window shaping, or manually trigger it at a time of their choosing.
```

### Kubernetes version upgrades

NAP node Kubernetes version upgrades behave differently from AKS managed agent pools. When the cluster control plane is upgraded to a new Kubernetes version, Karpenter recognizes the version delta soon after the update (subject to a polling interval) and marks affected nodes as drifted. Replacement is then driven by the standard disruption flow, respecting maintenance windows and disruption budgets, but customers cannot separately stage or defer the node k8s version upgrade the way they can with AKS agent pool upgrade controls.

Additionally, a Kubernetes version upgrade also triggers a node image version refresh: nodes are replaced with the effective node image set resolved for the new Kubernetes version. In practice this typically means moving to the latest compatible node image, subject to the same image-resolution and maintenance-window behavior described above. This means a control plane upgrade causes both a k8s version change and a node image change on NAP nodes simultaneously, neither of which is individually controllable today.

> ### Gap this design addresses

Author: TODO: remove this since it's covered by Goals section already

4. Provide a `kubernetesVersion` spec field that allows customers to specify the desired Kubernetes version for drift detection, decoupling node k8s version drift from the cluster control plane version.
5. Surface both the current node image version and the recently used node image version (with its paired Kubernetes version) in AKSNodeClass status.
6. Keep the API design extensible for future image version selectors.

## Non-Goals

1. Implement long-duration or arbitrary image pinning beyond the current and previously used versions.
2. Introduce prepared image spec by full resource ID.
3. Guarantee rollback to arbitrary historical versions.
4. Block provisioning scale-up while rollback state is being evaluated.

## API Changes

### API Field Grouping and Wrapper Name

`nodeImageVersion` and `kubernetesVersion` should be grouped under a dedicated sub-section rather than placed at the top level of spec. Top-level placement would crowd the spec namespace and leave no clean home for future image version selectors.

The proposed shape is:

```yaml
spec:

>   versionSelection:

Reviewer: I don't love `versionSelection` - I think because "selection" in this context is somewhat of a nothing-word.

Other top level fields are things like:
`localDNS`, `gpu`, `linuxOSConfig`, `kubelet`

Wondering if we just want to do `versions` here?

    kubernetesVersion: "1.32"
    nodeImageVersion: "202601.15.0"
    # future:

>     selectors:

Reviewer: If we had selectors I would think the syntax would be a slice of `SelectorTerm`'s, like [aws does](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/apis/v1/ec2nodeclass.go#L56).

Here's a report of how each selector works in AWS:
# EC2NodeClass selector behavior when multiple resources match

`EC2NodeClass` currently has five selector-style fields. They do **not** share one common "pick the first match" behavior.

| Field | Multiple matches |
|---|---|
| `amiSelectorTerms` | Selects the newest compatible, non-deprecated AMI for each compatibility set |
| `securityGroupSelectorTerms` | Attaches **all** matching security groups |
| `subnetSelectorTerms` | Keeps all matches, then selects the subnet with the most available IPs in each eligible AZ |
| `capacityReservationSelectorTerms` | Keeps all eligible reservations and models them as reserved-capacity offerings; launch-time filtering chooses among them |
| `placementGroupSelector` | Resolves exactly one placement group by exact name or ID |

## Common selector-term semantics

For `amiSelectorTerms`, `securityGroupSelectorTerms`, `subnetSelectorTerms`, and `capacityReservationSelectorTerms`:

- Conditions within one term are **ANDed**.
- Separate terms are **ORed**.
- Results matching multiple terms are de-duplicated by resource ID.
- Array order does not express preference.

For example, this means `(tag A AND tag B) OR explicit ID`; it does not mean "try the first term, then fall back to the second."

## `amiSelectorTerms`

All discovered AMIs are first divided by their compatibility requirements. Those requirements can include architecture and, for managed aliases, accelerator-related requirements.

For each equivalent requirement set, Karpenter retains one AMI using this order:

1. Prefer a **non-deprecated** AMI over a deprecated one.
2. If both have the same deprecation state, prefer the newest `CreationDate`.
3. If creation dates are equal, use the AMI name as a deterministic tie-breaker.

At provisioning time, Karpenter finds the retained AMI whose requirements are compatible with the candidate instance type. This means different nodes can use different selected AMIs—for example, ARM64 and AMD64 nodes can use separate AMIs.

If no AMI is compatible with the candidate instance types, provisioning fails.

One important limitation is that custom AMI discovery generally infers only architecture. Karpenter cannot infer that one custom AMI is GPU-specific, for example; separate `EC2NodeClass` resources should be used for distinctions Karpenter cannot represent.

Source: [AMI provider implementation](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/amifamily/ami.go)

## `securityGroupSelectorTerms`

Karpenter does **not** choose one security group. It:

1. Resolves every security group matching any term.
2. De-duplicates them by security-group ID.
3. Places all resolved IDs in the launch template.
4. Attaches all of them to the node's primary ENI and other applicable ENIs.

Consequently, an overly broad selector can attach several security groups and combine all their rules. It can also exceed AWS security-group-per-ENI limits or conflict with software that expects only one specially tagged security group, such as some AWS Load Balancer Controller configurations.

If no security groups match, `SecurityGroupsReady` becomes false and the `EC2NodeClass` is not ready, preventing provisioning.

Sources:

- [Security-group resolution](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/securitygroup/securitygroup.go)
- [Launch-template attachment](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/launchtemplate/launchtemplate.go)

## `subnetSelectorTerms`

All matching subnets are retained in `status.subnets`, but Karpenter does not send every subnet in an AZ to EC2.

Immediately before launch, Karpenter:

1. Groups matching subnets by Availability Zone.
2. Selects the subnet with the greatest estimated number of available IP addresses in each zone.
3. Accounts for IPs expected to be consumed by launches currently in flight.
4. Supplies one selected subnet per eligible zone to EC2 Fleet.
5. EC2 Fleet ultimately chooses the instance-type/AZ offering that fulfills the launch.

Therefore:

- **Within one AZ:** the subnet with the most available IPs wins.
- **Across AZs:** there can still be several eligible subnet/AZ choices, and the final AZ is determined as part of offering selection and EC2 Fleet fulfillment.
- If the `NodeClaim` is constrained to one zone, only that zone's selected subnet is usable.

An equal-IP-count tie has no separately documented subnet-ID tie-break; selector order should not be used to control it.

If no subnet matches, `SubnetsReady` becomes false and provisioning is blocked.

Source: [Subnet selection implementation](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/subnet/subnet.go)

## `capacityReservationSelectorTerms`

All matching, non-expired reservations are de-duplicated and exposed through `status.capacityReservations`. Each compatible reservation becomes a distinct reserved-capacity offering constrained by:

- Reservation ID
- Instance type
- Availability Zone
- Reservation type
- Interruptibility
- Available reservation capacity

There is no unconditional "newest reservation wins" rule. Selection happens through scheduling and launch-time filtering:

1. `reserved` capacity is preferred over `spot` and `on-demand` when the `NodePool` allows it.
2. Unavailable, empty, expiring, incompatible, or zonally shifted reservations are excluded.
3. Karpenter selects a compatible reservation-type partition because one Fleet request cannot mix every reservation type.
4. For multiple reservations for the same instance type and zone, Karpenter keeps the reservation with the **greatest available instance count**.
5. Capacity Blocks receive additional filtering because EC2 Fleet permits only one Capacity Block target in a request.
6. Karpenter creates reservation-specific launch templates and explicitly targets the chosen reservation ID.

If two same-pool reservations have equal available capacity, no stable reservation-ID tie-break is documented; selectors should be narrowed or the reservation-ID scheduling label should be used when a particular reservation is required.

Unlike AMIs, subnets, and security groups, an empty reservation match does not inherently make the `EC2NodeClass` unusable. If the `NodePool` also permits `spot` or `on-demand`, Karpenter can fall back. A workload restricted to `karpenter.sh/capacity-type: reserved` cannot launch without compatible available reservations.

Sources:

- [Capacity-reservation provider](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/capacityreservation/provider.go)
- [Reserved offering resolution](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instancetype/offering/reserved_capacity_resolver.go)
- [Launch-time reservation filtering](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/instance/filter/filter.go)

## `placementGroupSelector`

This differs from the term-list selectors:

- It accepts exactly one `name` or one `id`.
- Name and ID are mutually exclusive.
- Both are exact AWS lookup keys—not tag or wildcard searches.
- Placement-group names and IDs identify a single group in the applicable AWS scope.

Therefore, it should normally resolve zero or one placement group. There is no meaningful multi-match preference policy.

Internally, if AWS unexpectedly returned more than one result, the current provider uses the first result, but that is an implementation detail with no documented ordering guarantee and should not be treated as selection behavior.

If the placement group is missing or unavailable, `PlacementGroupReady` becomes false and the `EC2NodeClass` is not ready. If found, all nodes using the class target that placement group; compatible instance types are filtered according to the group's strategy.

Sources:

- [Placement-group lookup](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/providers/placementgroup/provider.go)
- [Placement-group readiness](https://github.com/aws/karpenter-provider-aws/blob/main/pkg/controllers/nodeclass/placementgroup.go)

## Summary

- **AMI:** newest compatible non-deprecated match.
- **Security group:** all matches.
- **Subnet:** most available IPs per AZ, then Fleet selects the launch offering.
- **Capacity reservation:** largest compatible reservation pool per instance-type/AZ after reservation-type filtering.
- **Placement group:** exact single name or ID; no normal multi-match case.

Reviewer: So it seems like if we did end up going with selectors how AWS does it:
1. It probably would be a list to match AWS.
2. List items OR'ed together, individual properties on items AND'ed, w/ CEL rules enforcing mutually incompatible stuff
3. The actual result of having a set of selectors that matched multiple versions (say) would have to be documented.

Translating that to how it _could_ initially work for versions:
```
nodeImageSelectors:
  - version: 202601.15.0
```

where `id` is the only possible field on `NodeImageSelectorTerm`, and we CEL-block multiple entries each with an ID (so basically must be length 1, to start?)

Same for Kubernetes version selector.

For **KubernetesVersion** I am not really sure there is any advantage to using a selector-like approach, because I suspect we will not allow multiple k8s versions within a single NodeClass... I guess the selectors could allow users to choose upgrade channels or something but that seems somewhat overly complicated and I tend to like the proposal of just:
1. Match AKS controlplane version (what we do today).
2. Pin to a fixed version (what you are proposing).

For **NodeImageVersion**, the small advantage that I see with selectors is that it works well if a single NodeClass has to support multiple images with different versions. I am not sure if that can ever happen in practice today (maybe no? I think you discuss it later), but it _could_ be worth hedging anyway in the API shape and then just forcing this shape:
```
spec: 
  versions:
    nodeImageSelectors:
    - version: 202601.15.0
```

since it's not all that much messier than:
```
spec: 
  versions:
    nodeImageVersion: 202601.15.0
```

and it's slightly more extensible.

Definitely a question I'd like @tallaxes thoughts on.

Author: re: the field name

i have a todo to just remove the "selectors" field from this section in general bc i want that to be punted to a future design. the important thing is that there is room for extensibility.

      validated: "true"
      ring: "stable"
```

> **Decision:** use `versionSelection`.

Author: TODO: rephrase this as "proposal" and remove the explicity "selectors" section and instead just say that future selectors could be nested within `versionSelection`

AKS enforces a version skew policy between the cluster control plane and node pools. Public AKS documentation is not perfectly internally consistent on the exact window: the supported versions FAQ says Kubernetes 1.28+ follows **three minor versions** of skew (N-3), while older agent pool REST descriptions still mention a tighter window. The design assumption here is that Karpenter must satisfy the AKS-supported skew window enforced by the platform and must not allow a node pool version greater than the control plane version.

When `spec.versionSelection.kubernetesVersion` is set, Karpenter must validate that the specified version satisfies this constraint relative to the cluster control plane version before provisioning or during drift evaluation:

1. The specified version must not be greater than the control plane version.
2. The specified version must remain within the AKS-supported minor-version skew window relative to the control plane version.

Violations should be surfaced as a condition on the NodeClass rather than silently accepted, since AKS is expected to reject machine creation once the requested version falls outside the supported skew window.

> **Version granularity:** v1 should accept **minor versions only** (for example `1.32`). This matches the simplest interpretation of AKS alias minor version behavior and avoids introducing patch-level drift semantics in the first iteration. When a customer specifies a minor version, Karpenter should treat that minor as the desired version and should not drift nodes solely because a newer patch within the same minor becomes available. Explicit patch-version support can be a follow-up design if customers need it.

Reviewer: I am not sure I follow this bit. Wouldn't the easiest thing be to just accept only a fully specified `x.y.z`? Does Machine support alias minor version? What if users want to have their machines on a uniform Kubernetes version? Specifying 1.32 isn't specific enough because then some machines will resolve to 1.32.1 and some to 1.32.2.

I am not convinced that default heterogeneity is what users want with this feature. My 2c is we should err in the other direction, just accept `x.y.z` and apply it exactly. If user wants to change it, they can change it to a newer `x.y.z` which will naturally trigger drift.

### nodeImageVersion Spec Field

The `nodeImageVersion` spec field is the unified customer surface for both node image pinning and rollback. It determines which node image release version Karpenter targets for new provisioning and drift evaluation.

**Semantics:**

1. **Unset (default):** Karpenter preserves the existing NAP image-resolution behavior. It resolves the latest gallery image version on every reconcile, updates `status.latestImageVersion`, and publishes the effective image set into `status.images` according to the current maintenance-window logic.

> 2. **Set to `status.latestImageVersion`:** Karpenter pins to the latest resolved version. New nodes are provisioned on that version. Automatic node image upgrades are paused — if a newer version becomes available, nodes will not drift until the customer updates or clears the pin.

Reviewer: ```suggestion
2. **Set to `status.latestImageVersion`:** Karpenter pins to the latest resolved version. New nodes are provisioned on that version, and existing nodes are drifted to that version. Automatic node image upgrades are paused — if a newer version becomes available, nodes will not drift until the customer updates or clears the pin.
```

Reviewer: right?

3. **Set to a value in `status.recentlyUsedVersions[*].imageVersion`:** Karpenter rolls back to that previously used version. The rollback validation rules from `status.recentlyUsedVersions` apply: the requested version must match an entry in the array and the Kubernetes version of that entry must be compatible.
4. **Set to any other value:** CEL admission validation rejects the request. The only valid values are `status.latestImageVersion` (pinning at latest) or a value present in `status.recentlyUsedVersions[*].imageVersion` (rollback to a previous version).

The latest image version is always visible through `status.latestImageVersion`. The previously used version is visible through `status.recentlyUsedVersions`.

**Decision: explicit version string.**

Customers set `nodeImageVersion` to the exact node image release version suffix they want Karpenter to target:

> ```yaml

Author: TODO: fix indentation here

**Alternative considered: boolean rollback flag**

A boolean field (`rollbackToPrevious: true`) was considered, which would let Karpenter automatically select the most recent entry in `status.recentlyUsedVersions` without the customer specifying it. This was rejected because:

1. It does not extend to pinning-at-current, where there is no "previous" to roll back to.
2. A customer could silently roll back to a version they did not intend.
3. If Karpenter later stores multiple previous image versions, a boolean becomes ambiguous without additional API to specify which entry to use.
4. The explicit version string approach mirrors AKS RP semantics, where `nodeImageVersion` is always a concrete version value.

### AKSNodeClass status: latestImageVersion and recentlyUsedVersions

Add a new status section that mirrors the AKS RP recently-used rollback model:

```go
// RecentlyUsedVersion records a node image version that was previously active
// on this NodeClass, used to validate rollback targets.
type RecentlyUsedVersion struct {
	// timestampUsed is the time this node image version was last active,
	// recorded when Karpenter moves status.images forward.
	// +optional

> 	TimestampUsed metav1.Time `json:"timestampUsed,omitempty"`

Reviewer: wdyt about
```suggestion
	LastUsed metav1.Time `json:"lastUsed,omitempty"`
```

instead of `timestampUsed`?

Author: I need to double check, but I think I grabbed this naming directly from Agentpool API. Other than the general desire to keep it as close to that as possible, I don't really mind this change. And I think `LastUsed` does make the meaning clearer.

Author: Yes, it's from agentpool API and it's what that returns. I'm still open to switching to `LastUsed` tho

	// imageVersion is the AKS node image release version suffix,
	// e.g. "202601.15.0".
	// +optional
	ImageVersion string `json:"imageVersion,omitempty"`

	// kubernetesVersion is the Kubernetes version paired with this image,

> 	// e.g. "1.32.5".

Reviewer: flagging another reason to think about https://github.com/Azure/karpenter-provider-azure/pull/1756/changes#r3725163832 - if this is the `x.y.z` version but we only accept `x.y` how do they roll back to this version? OTOH, if we only show `x.y` here it seems it might be possible to end up with a patch version that's newer than the associated `ImageVersion`, which I thought we wanted to avoid (the reason we maintain the AKS behavior that k8s upgrade forces image upgrade)

Author: Yeah I have this to wrap my head around a bit more (with a dive into RP code/API too). I think I agree with your other comment saying let's just ensure that cx has to specify the fully-qualified patch version, but also want to make sure I think through the implications.

Author: Decided to require explicit `x.y.z` for now for simplicity. If customers set their k8s version, then they opt out of automatic k8s version drifting to newer patch/minor versions.

4. If valid, set the effective target image release version suffix to the matched `recentlyUsedVersions` entry's `imageVersion`.
5. Apply the rollback image version suffix across the resolved image definitions, publish any matching subset into `status.images`, and surface a warning if some compatible definitions do not have that suffix.

### Image selection during rollback

status.images is already a list because a single AKSNodeClass can expose multiple compatible image definitions. For example, the same NodeClass may publish Gen2 amd64, Gen1 amd64, and Gen2 arm64 images, each with requirements that determine which instance types can use it.

Rollback should preserve this model:

1. Karpenter resolves the normal goal image list from the AKSNodeClass, Kubernetes version, FIPS mode, SIG/CIG mode, and supported image definitions.
2. Each resolved image keeps its existing requirements. Node launch continues to choose the first image whose requirements are compatible with the selected instance type.

> 3. Rollback does not choose one image from status.images globally. Instead, it rewrites every resolved image ID to the requested release version suffix.

Author: TODO: rephrase/update the info here, because rollback should be checking for image presence in the gallery + then filtering for specific version suffixes, and only returning the image IDs that are available for the rollback version.

```

The same suffix rewrite is applied independently to each resolved image definition. This lets multiple NodePools share one AKSNodeClass while still rolling back to the image variant selected by each NodePool's instance type requirements.

Before using rolled-back images, Karpenter should verify which resolved image definitions actually have the requested release suffix.

1. Definitions that do not have the requested suffix are omitted from `status.images`, while definitions that do have it remain publishable.
2. Karpenter surfaces a warning condition so operators know that some compatible variants are temporarily unavailable for provisioning.
3. If no resolved image definitions have the requested suffix, Karpenter surfaces a failure condition and does not publish rollback-effective `status.images`.

### Implementation options for applying rollback

Two options exist for where to apply the version suffix rewrite:

- **Option 1 — Keep `status.images` as latest, convert at use sites:** `status.images` always holds the latest resolved images. A helper (e.g. `convertToRolledBackImage`) rewrites the version suffix at each consumer call site (provisioning, drift).

> - **Option 2 — Write effective images into `status.images`:** The reconciler applies the version suffix rewrite before publishing `status.images`. All consumers automatically receive the correct images with no call-site changes. A separate `status.latestImageVersion` field always tracks the latest gallery version regardless of the active pin.

Reviewer: I am a bit confused as to why this is a "version suffix rewrite".

Is the issue that the actual AgentBaker API doesn't return multiple versions currently, it only returns the latest, so we can't just call the API with a filter like `version=<rollbackVersion>` and actually get/validate the versions from the real API response the same way we do for pin-to-latest?

If so, it may be worth filing a specific Github item to track improving that, as it should be possible I would think.

Reviewer: OK with suffix rewrite as a hack for now though it should be safe enough unless somebody is mucking with status manually.

Author: added a todo above to rephrase/edit this, because I think it was left over from an earlier idea and it also was surprising to me when re-reading the doc fully this morning.

https://github.com/Azure/karpenter-provider-azure/pull/1756/changes#r3723560185

`status.latestImageVersion` is always updated to the latest gallery version regardless of the active pin, serving two purposes: (1) it is the valid "pin at current" target for CEL validation, and (2) it lets operators see whether a newer version is available while the cluster is pinned. It is intentionally a gallery-view field, not a guarantee that the same version is currently effective in `status.images`.

Option 1 was rejected because a missed call site would silently provision the wrong image — Option 2 is safer by default.

## Validation and Conditions

Rollback validation rejects a request when `recentlyUsedVersions` is empty, or when the desired Kubernetes version is incompatible with the matched entry's `kubernetesVersion`. If admission-time validation cannot cleanly evaluate status fields, the reconciler rejects the request via condition.

> **Proposed conditions:**

Author: TODO: review conditions. I don't love "Ignored". Maybe we should say Failed? Or NotAllowed?

> | Condition type | Reason | Meaning |

Reviewer: Can you clarify which of these conditions roll up to impact the overall AKSNodeClass ready condition and which do not?

Author: Added this context to the doc

|---|---|---|

> | `ImageRollbackReady` | `RecentlyUsedVersionsNotAvailable` | No valid rollback target exists |

Reviewer: Do we actually need all of these different condition types? Today we have 
```
const (
	ConditionTypeImagesReady            = "ImagesReady"
	ConditionTypeKubernetesVersionReady = "KubernetesVersionReady"
	ConditionTypeSubnetsReady           = "SubnetsReady"
	ConditionTypeLocalDNSReady          = "LocalDNSReady"
)
```

You are proposing adding 4 more, so half of the overall conditions we have would be about pinning in that case.

We already have:
`ConditionTypeImagesReady`
and
`ConditionTypeKubernetesVersionReady`

Which I feel like work well for things like:
```
| `ImageRollbackActive` | `KubernetesVersionMismatch` | k8s version incompatible with rollback target |
| `NodeImageVersionPinned` | `ImageVersionUnavailable` | None of the resolved image definitions have the requested suffix, so no effective image set can be published |
```

So above could instead just be `ConditionTypeImagesReady/KubernetesVersionMismatch` + `ConditionTypeImagesReady/ImageVersionUnavailable`

Then we definitely need another condition for "warning" like `ImageVersionPartiallyAvailable`.

Do we actually need conditions for the other cases? I would assume any rollback we couldn't apply would be an error in the images condition? The other conditions like "Rollback Applied" are implied by images ready + kubernetes version ready, are they not?

Author: yeah I read through the doc again this morning + think this list is overkill. have a todo to rework a bit. will also include clear error vs. warning conditions
https://github.com/Azure/karpenter-provider-azure/pull/1756/changes#r3723580102

Author: Did an overhaul of this section. 

For now, I'm gonna leave the "partially available" warning out, as well. This is for two reasons: 1) do customers actually care if there are fewer variants available to be selected by Karpenter? and 2) I don't think we need to add additional logic right now to determine if the set of variants is smaller _due to us setting node image version_ or due to some other constraint

| `NodeImageVersionPinned` | `ImageVersionUnavailable` | None of the resolved image definitions have the requested suffix, so no effective image set can be published |
| `NodeImageVersionPinned` | `ImageVersionPinnedAtCurrent` | Pinned to latest; auto-upgrades paused |
| `NodeImageVersionPinned` | `ImageVersionPinnedAtPrevious` | Pinned to previous version; rollback active |
| `KubernetesVersionControlled` | `KubernetesVersionMismatch` | Node k8s version differs from `spec.versionSelection.kubernetesVersion` |

### nodeImageVersion and kubernetesVersion Validation

In addition to rollback-specific validation, the following rules apply to `nodeImageVersion` and `kubernetesVersion`:

**CEL admission validation for `nodeImageVersion`:**

`spec.versionSelection.nodeImageVersion` must equal `status.latestImageVersion` (pin at current) or one of the `imageVersion` values in `status.recentlyUsedVersions` (rollback). All other values are rejected. Note: status-dependent CEL rules may require a webhook validator; reconcile-time validation with a clear condition is acceptable as a fallback.

> If the resolved image definitions for one NodeClass do not all have the requested suffix available, Karpenter accepts the request, publishes the subset of matching image definitions into `status.images`, omits the non-matching definitions, and surfaces a warning condition. If no matching image definitions remain, Karpenter surfaces a failure condition and does not publish rollback/pin-effective `status.images` for that request.

Author: TODO: verify if we have access at some point to the list of variants we _expect_ to have, or if we just know at the end of the filtering/listing process if we have > 1 image ready.

### Kubernetes version drift

When `spec.versionSelection.kubernetesVersion` is set, Karpenter uses it as the desired k8s version for drift detection instead of the cluster control plane version. Nodes running a different version are drifted and replaced. New nodes are provisioned with the image compatible with the specified version. If unset, drift falls back to comparing against the control plane's current version. `kubernetesVersion` participates in AKSNodeClass hashing, so changing it triggers NodeClassDrift.

## Decision Notes

### Decision 1: Snapshot storage in status

Conclusion: Use AKSNodeClass status.recentlyUsedVersions.

Rationale:

1. etcd-backed durability across operator restarts.
2. Closest coupling to current status.images lifecycle.
3. Mirrors the AKS RP rollback allowlist shape.
4. No new external state store required.

### Decision 2: Array-based version history

Conclusion: Use `[]RecentlyUsedVersion` (array) rather than a single pointer.

Rationale:

1. Matches the AKS RP preview API shape (`recentlyUsedVersions` is an array there too).

> 2. Enables multi-environment staged rollout scenarios by retaining more than one historical version.

Author: TODO: clarify that this allows _future_ retention of multiple versions

3. The maximum number of entries to retain is an open question; a single entry would match current AKS RP behavior, while more entries extend rollback reach.

Future consideration: the API must define how rollback selects the target entry when multiple are present — most recent match, explicit index, or another mechanism. See Out of Scope Follow-up Design item 5.

## Open Questions

> 1. Do we need an admission-time validation webhook for rollback request semantics, or is reconcile-time validation sufficient? This is particularly relevant for `nodeImageVersion` validation against status fields, which may not be accessible to standard CRD CEL validators at admission time.

Reviewer: I think status validation is fine, we use it a lot elsewhere already.

> 2. Should rollback support only the AKS Machine API path, or should it explicitly support both AKS Machine API and the node bootstrapping client/VM path? Current expectation is that it should work either way because both paths consume status.images, but this should be verified.

Reviewer: I think don't couple it to machine if we don't have to.

3. Does the existing node image cache require rollback-specific invalidation or cache-key changes so that rollback requests and roll-forward after rollback are reflected immediately?

> 4. When auto-upgrade or a future image policy moves the pool forward after rollback, should the rollback request be cleared, or should it remain set and become ignored/invalid?

Reviewer: auto upgrade doesn't impact Karpenter. IMO if user rolls back in spec, it's pinned til they un-roll it back.

> 5. In a future follow-up, should `kubernetesVersion` also accept full patch versions (e.g. `1.32.5`), and if so should that mean exact patch pinning or another semantics?

Reviewer: See previous comments, I think it should start accepting full patch.

4. Support-window or stale-image warnings for long-duration pinned images.
5. **Multi-environment staged rollout across clusters.** The current design stores only one `recentlyUsedVersions` entry per NodeClass. This means a cluster can only target the current latest or the immediately preceding image version. For staged rollout pipelines where validation takes longer than one image release cycle — for example, dev validates `202607.15.0` while `202608.15.0` and `202609.15.0` have already shipped — downstream clusters can no longer pin to the validated version because it has been evicted from the single-entry history. Supporting reliable multi-cluster staged rollout will require storing multiple `recentlyUsedVersions` entries (a list of the last N versions) rather than a single pointer, so that a validated version remains reachable for pinning even if several newer versions have since been released.

> ## Production Readiness

Author: TODO: remove this section

## Overview

This design adds two user controls to NAP NodeClasses:

1. `spec.versions.nodeImageVersion` lets a customer pin or roll back the node image version.
2. `spec.versions.kubernetesVersion` lets a customer control the node Kubernetes version used for drift and provisioning new nodes.

The main use cases are:

- pause automatic node image roll-forward
- roll back to a previously used image version after a bad image release
- decouple node Kubernetes version drift from the control plane version

The design does **not** introduce arbitrary image IDs or arbitrary historical rollback. It stays within status-backed values already visible on the NodeClass.

## Current Problem

Today, NAP continuously resolves the latest compatible gallery images and publishes them into `status.images` using existing maintenance-window behavior.

That creates two gaps:

1. Customers cannot explicitly hold the current image version or roll back to a previously used one.

> 2. Customers cannot independently choose the node Kubernetes version used for drift; upgrades effectively track control-plane changes.

Reviewer: minor, phrasing: 
```suggestion
2. Customers cannot independently choose the Kubernetes version installed on each node (used for drift); upgrades effectively track control-plane changes.
```

Not sure if this is better though so if you prefer current phrasing feel free to resolve this and not change.

The result is that a control-plane upgrade can drive both a Kubernetes version move and a node image move at the same time, with little direct operator control.

## Goals

> 1. Allow pinning and rollback through a single `nodeImageVersion` field.

Reviewer: ```suggestion
1. Allow pinning and rollback of node image version through a single `nodeImageVersion` field.
```

> 1. Allow pinning and rollback through a single `nodeImageVersion` field.

Reviewer: minor: Worth calling out explicitly that a goal is "let users do with Karpenter/NAP what they can already do with AKS" -- we're scoping support to basically what AKS already allows, which informs a lot of these other requirements because those are just making explicit what AKS already allows.

Is that right?

> 2. Preserve last effective image/Kubernetes version pairs in NodeClass status.

Reviewer: ```suggestion
2. Preserve last effective image/Kubernetes version pairs in NodeClass status, for use in rollbacks
```

> 3. Keep the API narrow by validating only status-backed version choices.

Reviewer: Not 100% sure I understand what this means, can you clarify a bit?

4. Allow an explicit `kubernetesVersion` for node drift and replacement decisions.
5. Leave room for future image selector-based controls.

## Non-Goals

1. Arbitrary long-term image pinning to any historical version.

> 2. Prepared image support by full resource ID.

Reviewer: ```suggestion
2. Prepared image spec (PIS) support by full resource ID.
```

> 3. Kubernetes version rollback.

Reviewer: We actually allow/support this? Or no? For some reason I thought we had to support this for some node image rollbacks, am I misremembering?

4. Global scale-up blocking beyond requests that resolve to an invalid target.

## Operating Model

> There are three version concepts in this design.

Reviewer: Clarify we're talking about kubernetesVersion here? I think there are more than 3 concepts otherwise. (Since there are some node image version concepts too?)

| Concept | Source | Meaning |
|---|---|---|
| Desired Kubernetes version | `spec.versions.kubernetesVersion` when set; otherwise `status.controlPlaneKubernetesVersion` | Version the NodeClass is trying to use for nodes |
| Effective node Kubernetes version | `status.kubernetesVersion` | Last accepted node Kubernetes version used for provisioning, image resolution, and drift |
| Observed control plane Kubernetes version | `status.controlPlaneKubernetesVersion` | Latest control-plane version observed by the reconciler |

For Kubernetes versions, the design uses two status fields.

| Field | Meaning |
|---|---|

> | `status.kubernetesVersion` | Effective node Kubernetes version currently used by provisioning, image resolution, and drift |

Reviewer: Note to future self to understand why we need this field if we have the spec field.

Reviewer: Ahh because we want to use the status as the source of truth for what we actually do.

| `status.controlPlaneKubernetesVersion` | Latest observed control-plane version used for compatibility checks and fallback behavior |

For node images, the design uses three status views.

| Field | Meaning |
|---|---|
| `status.latestImageVersion` | Latest gallery version known to the reconciler |

> | `status.images` | Effective image set currently used for provisioning and drift |

Reviewer: minor: clarify this is existing? Or maybe clarify for each of these fields (and above k8s image version ones) which is existing and which is new?

    // +optional
    NodeImageVersion *string `json:"nodeImageVersion,omitempty"`
}

type AKSNodeClassSpec struct {
    // versions controls the Kubernetes and node image versions for the NodeClass.
    // If omitted, both versions follow their automatic defaults.
    // +optional
    Versions *Versions `json:"versions,omitempty"`
}
```

The `versions` wrapper keeps the v1 surface narrow and leaves a natural home for future selectors such as `spec.versions.nodeImageSelectors`.

### Status

```go
type RecentlyUsedVersion struct {
    // timestampUsed is when this image version was last effective.
    // +optional
    TimestampUsed metav1.Time `json:"timestampUsed,omitempty"`
    // imageVersion is the node image version suffix.
    // +optional
    ImageVersion string `json:"imageVersion,omitempty"`
    // kubernetesVersion is the Kubernetes version paired with the image version.
    // +optional
    KubernetesVersion string `json:"kubernetesVersion,omitempty"`
}

type AKSNodeClassStatus struct {
    // images contains the current set of images available to use for the NodeClass.
    // +optional

>     Images []NodeImage `json:"images,omitempty"`

Reviewer: I think images is already here, but the other fields proposed are all new, correct?

Wondering if they also should be in a subobject in status like `versions` just for slightly better structure?

Reviewer: Proposed versions in status too... we can debate about this more on the actual PR containing the API changes as well as I am not married to any of this atm.

    // kubernetesVersion contains the current kubernetes version which should be used for nodes provisioned for the NodeClass.
    // +optional
    KubernetesVersion *string `json:"kubernetesVersion,omitempty"`
    // controlPlaneKubernetesVersion is the latest observed control plane version.
    // +optional
    ControlPlaneKubernetesVersion *string `json:"controlPlaneKubernetesVersion,omitempty"`
    // latestImageVersion is the latest node image version resolved from the gallery.
    // +optional
    LatestImageVersion string `json:"latestImageVersion,omitempty"`
    // recentlyUsedVersions contains previously effective node image versions.
    // +optional
    RecentlyUsedVersions []RecentlyUsedVersion `json:"recentlyUsedVersions,omitempty"`
}
```

## Kubernetes Version Semantics

### Default behavior

If `spec.versions.kubernetesVersion` is unset:

1. reconcile observes the control plane version
2. reconcile validates it
3. reconcile publishes it into `status.kubernetesVersion`
4. drift and provisioning use `status.kubernetesVersion`

> This preserves the existing behavior while making the effective node version explicit in status.

Reviewer: ```suggestion
This preserves the existing behavior while making the effective Kubernetes version explicit in status.
```

maybe? Node is a bit ambiguous because node version could be for k8s or image

The requested version must satisfy AKS node/control-plane skew rules.

Notation:

- `C = cMajor.cMinor.cPatch` is the control-plane version
- `N = nMajor.nMinor.nPatch` is the requested node version

1. major versions must match
    (`nMajor == cMajor`)
2. node minor must not be greater than control-plane minor
    (`nMinor <= cMinor`)
3. node minor must be at most three minors behind control-plane minor
    (`cMinor - nMinor <= 3`)
4. when both are on the same minor, node patch must not be greater than control-plane patch
    (`nMinor == cMinor -> nPatch <= cPatch`)
5. the requested version must not be lower than the current effective node Kubernetes version
    (if `E = eMajor.eMinor.ePatch` is the current effective node version, then `N >= E`)

This design does not support Kubernetes version rollback.

## Node Image Version Semantics

`spec.versions.nodeImageVersion` is the single customer control for pinning and rollback.

Valid values are only:

1. the current effective image version from `status.images[]`
2. the latest resolved image version from `status.latestImageVersion`
3. a previously effective version from `status.recentlyUsedVersions[*].imageVersion`

> `nodeImageVersion` is only valid when `spec.versions.kubernetesVersion` is also set. Anything else is invalid.

Reviewer: minor: May be worth explaining slightly _why_ this is? I think it's because we want to maintain the invariant that a _new_ Kubernetes version cannot be run on an image version published before that k8s version existed. AKS RP does not allow this and does not test for it, so it may not work.

### Default behavior

If `spec.versions.nodeImageVersion` is unset:

- NAP continues its existing automatic image-selection behavior
- `status.latestImageVersion` keeps tracking the newest gallery version
- `status.images` remains governed by existing maintenance-window behavior for automatic gallery updates

### Pinning behavior

If `spec.versions.nodeImageVersion` equals:

- the current effective image version: pin to current
- the current `status.latestImageVersion`: pin to that specific latest value at update time

In both cases, automatic image roll-forward pauses until the customer changes or clears the pin.

### Rollback behavior

If `spec.versions.nodeImageVersion` matches an entry in `status.recentlyUsedVersions[*].imageVersion`, the request is a rollback request.

Rollback is allowed only when the selected historical entry is **rollback-compatible**.

**Rollback-compatible Kubernetes version** means:

> - the current desired Kubernetes version exactly equals the selected `status.recentlyUsedVersions[*].kubernetesVersion`

Reviewer: 🤖 I think rule 5 is the issue here: it forbids lowering the effective **node** Kubernetes version, even though this rollback does not require a control-plane downgrade. For example, after the control plane and nodes move from `1.32.5` to `1.32.7`, selecting the historical image pair requires moving the nodes back to `1.32.5` while leaving the control plane at `1.32.7`. If AKS allows node downgrades that remain within supported skew, this rule unnecessarily makes the rollback target unusable. Please clarify node-downgrade semantics separately from control-plane downgrade and remove or relax this prohibition if supported.

This is exact equality, not a looser skew-compatible rule.

## Status Semantics

### Kubernetes status fields

> | Field | Meaning |

Reviewer: minor: this table seems somewhat duplicated, compared to section "For Kubernetes versions, the design uses two status fields." above.

|---|---|
| `status.kubernetesVersion` | Effective node Kubernetes version |
| `status.controlPlaneKubernetesVersion` | Latest observed control-plane version |

Rules:

1. `status.kubernetesVersion` is the source of truth for provisioning, image resolution, and drift.
2. `status.controlPlaneKubernetesVersion` is used for skew validation and observability.
3. A control-plane refresh always updates `status.controlPlaneKubernetesVersion`, even if the effective node version remains pinned or becomes incompatible.

### Image status fields

| Field | Meaning |
|---|---|

> | `status.images` | Effective image set |

Reviewer: Same comment as above, seems somewhat duplicate compared to ln ~75 table?

## Reconciliation Model

### Snapshot trigger

A snapshot into `status.recentlyUsedVersions` is taken whenever the effective image set moves to a different image version suffix.

Triggers:

1. automatic gallery advance becomes effective
2. customer sets `nodeImageVersion`
3. customer unsets `nodeImageVersion`
4. effective Kubernetes version change produces a different image set

### Publication ordering

When reconcile is about to move the effective node Kubernetes version and/or effective image set, it must:

1. validate the request
2. resolve the goal image set
3. snapshot the old effective pair from `status.kubernetesVersion` and `status.images`
4. publish the new effective pair

This ensures `recentlyUsedVersions[*].kubernetesVersion` records the Kubernetes version with which the historical image was actually used.

## Explicit Image Version Changes

This flow applies to both rollback and pinning, and only when `spec.versions.kubernetesVersion` is also set.

1. validate that `spec.versions.kubernetesVersion` is set
2. validate the requested `nodeImageVersion` against status-backed choices
3. if it is a rollback target, validate rollback-compatible Kubernetes version equality

> 4. resolve concrete images through the normal `List()` flow

Reviewer: 🤖 The normal `List()` flow only returns the latest version per image definition. Filtering that result to a historical `nodeImageVersion` will therefore leave zero images for a rollback target and make the NodeClass unready. The earlier suffix-rewrite mechanism handled this, but it was removed in the rewrite. Please define how historical SIG/AgentBaker and CIG image IDs are resolved before filtering.

## Automatic Gallery Updates

Automatic updates keep the existing maintenance-window behavior.

That means:

1. reconcile may discover a newer gallery version and publish it into `status.latestImageVersion`
2. `status.images` may remain on the previous effective set while the maintenance window is closed
3. once the maintenance window opens, reconcile may publish the newer effective set into `status.images`

This preserves the current separation between “latest known” and “currently effective.”

## Validation Strategy

### Summary by phase

| Phase | Purpose |
|---|---|
| Admission | Best-effort format, transition, and persisted-state validation |
| Reconcile on spec change | Fresh lookups, authoritative compatibility checks, image-resolution validation |
| Periodic reconcile | Refresh observed control plane version and re-check accepted effective state |

### `nodeImageVersion` checks

| Rule | CEL admission | Reconcile failure |
|---|---|---|
| `spec.versions.kubernetesVersion` is set when `spec.versions.nodeImageVersion` is set | Reject | `ValidationSucceeded=False`, `NodeImageVersionRequiresKubernetesVersion` |

> | Requested value must match a version already exposed in NodeClass status | Best-effort reject | `ValidationSucceeded=False`, `NodeImageVersionInvalid` |

Reviewer: What is a Best-effort reject? This seems like just a standard reject?

Reviewer: Ahh "CEL is best-effort because it only sees persisted state." - probably not worth calling that out here just say reject?

| Rollback target is rollback-compatible | Best-effort when persisted state is sufficient | `ValidationSucceeded=False`, `RollbackTargetKubernetesVersionMismatch` |
| Requested version resolves at least one concrete image | Not checked | `ImagesReady=False`, `RequestedNodeImageVersionUnavailable` |

### `kubernetesVersion` checks

| Rule | CEL admission | Reconcile failure |
|---|---|---|

> | Value is full `major.minor.patch` | Reject | `ValidationSucceeded=False`, `KubernetesVersionInvalidFormat` |

Reviewer: I thought we changed the design to suggest that we always require full major.minor.patch?

| Requested patch exists in AKS-supported metadata | Not checked | `KubernetesVersionReady=False`, `KubernetesVersionUnsupported` |

Notes:

- CEL is best-effort because it only sees persisted state.
- Reconcile is authoritative because it performs fresh control-plane and supported-version lookups.
- Transient lookup or provider failures should return an error for retry rather than set a sticky `False` condition.

## Conditions and Reasons

The design reuses existing readiness-bearing conditions.

| Condition | Meaning when false |
|---|---|
| `ValidationSucceeded` | Requested spec is invalid for current object state |
| `KubernetesVersionReady` | Effective Kubernetes version is unsupported or incompatible |
| `ImagesReady` | No concrete images can be resolved for the effective target |

Recommended new reasons:

| Condition | Reason | Meaning |
|---|---|---|
| `ValidationSucceeded=False` | `NodeImageVersionInvalid` | Requested `nodeImageVersion` does not match a version already exposed in NodeClass status |

> | `ValidationSucceeded=False` | `NodeImageVersionRequiresKubernetesVersion` | `nodeImageVersion` is set while `kubernetesVersion` is unset |

Reviewer: I am not sure we actually need a rule for this because we can block at the CEL level can't we?

| `ValidationSucceeded=False` | `RollbackTargetKubernetesVersionMismatch` | Requested rollback target is not rollback-compatible |

> | `ValidationSucceeded=False` | `KubernetesVersionDowngradeNotSupported` | Requested Kubernetes version is a downgrade |

Reviewer: I don't think we actually need/want to block this, downgrade is supported as part of rollback (indeed it is _required_ as part of rollback AFAIK)

| `KubernetesVersionReady=False` | `KubernetesVersionUnsupported` | Requested patch is not supported by AKS metadata |
| `KubernetesVersionReady=False` | `KubernetesVersionControlPlaneIncompatible` | Requested version violates control-plane compatibility |
| `ImagesReady=False` | `RequestedNodeImageVersionUnavailable` | Requested image version resolved zero concrete images |

## Drift and Provisioning

The governing rule is simple:

- drift and provisioning always use the published effective status fields

### Provisioning

New NodeClaims use:

- `status.kubernetesVersion` for the effective node Kubernetes version
- `status.images` for the effective image set

New NodeClaims are only launched after the NodeClass is `Ready` for the current generation. This ensures launch does not consume stale effective status while reconcile is still validating or publishing a spec change.

If filtering resolved images to the requested `nodeImageVersion` leaves no concrete images, Karpenter cannot provision new NodeClaims from that NodeClass.

### Drift

A spec change can make an existing NodeClaim drifted immediately, but that does not by itself mean replacement is actionable.

> 1. `spec.versions` fields participate in the AKSNodeClass hash, so their changes enqueue affected NodeClaims.

Reviewer: I think we will want to exclude this field from the hash, because it doesn't actually control what is provisioned, it just controls what gets put into status, and that status field controls what the nodes actually get.

This also means that we will NOT drift nodes that the user pins to the version they are on, which is good.

Reviewer: Actually wait I am wrong, we need to exclude `spec.versions.nodeImageVersion` from drift, but not `spec.versions.kubernetesVersion`.

Reviewer: Hmm actually my wrongness is wrong, I think we want to exclude the whole field.

- when both `spec.versions.kubernetesVersion` and `spec.versions.nodeImageVersion` are set, drift follows the accepted explicit pair
- when it is unset, drift follows the effective version published from `status.controlPlaneKubernetesVersion`

Replacement must not proceed until the NodeClass is `Ready` for the current generation and the effective status for that generation has been published.

If the latest observed control plane version and the effective node Kubernetes version become incompatible, drift should not try to converge toward an invalid target. The NodeClass should remain unready until the compatibility issue is resolved.

## Decisions

### Decision 1: Use status for rollback history

Store rollback history on the AKSNodeClass itself.

Rationale:

1. durable across operator restarts
2. tightly coupled to the effective status lifecycle
3. mirrors the AKS RP recently-used rollback shape
4. avoids introducing another state store

### Decision 2: Use an array for history

Store `recentlyUsedVersions` as an array, not a single entry.

Rationale:

1. matches the AKS RP preview shape

> 2. preserves more than one rollback candidate

Reviewer: minor: let's just keep 1 item in the history for now, same as AKS
